feat(Multiple Languages): Show ProjectLanguageChooser in AT

hirokiterashima opened the pull request:
## Note
- Don't worry about styles in this PR
- Remember to use the ```[field].i18n = { id: "abc123", "modified": 123 }``` project format for this PR and all future PR's.

## Changes
- Change ProjectLanguageChooserComponent so we can use it in the VLE and the AT
   - Take in ```ProjectLocale``` as ```@Input``` 
   - Emit selected ```Language``` as ```@Output```
   - Move to common folder
- Display ProjectLanguageChooserComponent in the AT's TopBarComponent if unit has translations
   - Update available language options when author changes them in the Project Info view

## Test AT
- ProjectLanguageChooserComponent displays supported languages iff unit has translations
- Editing language options in the Project Info view updates the ProjectLanguageChooserComponent's available options
   - If there are no available languages, the ProjectLanguageChooserComponent should not appear 

*Note: nothing should happen we switch the language option in the ProjectLanguageChooserComponent atm. We'll handle that in a future PR.

## Test Student VLE
- Switching between translations works as before. While logged in as a student or previewing a unit, test that:
   - Languages drop-down appears only if the unit has at least 1 supported locale
   - Switching between the language options immediately translates elements of the page that have translations
      - Test everywhere: unit plan, step select drop down, unit title
      - Elements that don't have translations appear in default locale
      - You can switch back to default locale

hirokiterashima committed:
Change ProjectLanguageChooser to take in ProjectLocale as input and
emit selected Language as output. This will let us reuse this component
in the AT.

hirokiterashima committed:
Move ProjectLanguageChooser to common folder and show it in the AT's
TopBarComponent. Update options when author changes them in the Project
info view

geoffreykwan reviewed (changes requested):
Mostly works but I found one issue.

In the Authoring Tool, if you have set the language in the top bar to one of the additional languages and then go to the Unit Info view and change the default language or add/delete an additional language, the language in the top bar changes back to the default language.

I think how we should handle this is if the author is in the Unit Info view and changes the default language or additional languages, the top bar language should not change. The top bar language should only change to the default language if the language that is currently selected in the top bar is removed from the additional languages.

hirokiterashima committed:
Set selected language to default language only when previously-selected
option is no longer available (i.e., removed by author), or this is the
first time the component is displayed

hirokiterashima commented:
@geoffreykwan Good catch. We now set selected language to the default language only when the previously-selected option is no longer available (i.e., removed by author), or this is the first time that the component is displayed. PTAL.

geoffreykwan commented on src/app/common/project-language-chooser/project-language-chooser.component.ts:
```diff
@@ -0,0 +1,36 @@
+    ) {
+      this.selectedLanguage = this.availableLanguages.find((lang) =>
+        this.projectLocale.isDefaultLocale(lang.locale)
+      );
```
Can this be changed to 
```
this.selectedLanguage = this.projectLocale.getDefaultLanguage();
```
now that there is a isSameLanguage() function that is used in [compareWith]?

hirokiterashima replied:
Yes, this works. I changed it.

geoffreykwan commented on src/app/common/project-language-chooser/project-language-chooser.component.spec.ts:
```diff
@@ -0,0 +1,57 @@
+    const newOptions = await getOptions();
+    expect(await newOptions[3].isSelected()).toBeTrue();
+  });
+});
```
Instead of 
```
await options[1].click();
```
you could do
```
const selectHarness = await loader.getHarness(MatSelectHarness);
await selectHarness.clickOptions({ text: 'Japanese' });
```
and then you can get the selected value by doing
```
expect(await selectHarness.getValueText()).toEqual('Japanese');
```

hirokiterashima replied:
Good suggestion. Made the code easier to understand.

geoffreykwan commented on src/app/common/project-language-chooser/project-language-chooser.component.spec.ts:
```diff
@@ -0,0 +1,57 @@
+  const selectComponent = selects[0];
+  await selectComponent.open();
+  return await selectComponent.getOptions();
+}
```
You can do this to get the first MatSelectHarness without having to get all of them and then getting the first one from the array.
```
const selectHarness = await loader.getHarness(MatSelectHarness);
```

hirokiterashima replied:
Good suggestion. Made the code easier to understand.

geoffreykwan commented on src/app/common/project-language-chooser/project-language-chooser.component.spec.ts:
```diff
@@ -0,0 +1,57 @@
+    setProjectLocale(new ProjectLocale({ default: 'en_US', supported: ['ja', 'es'] }));
+  });
+
+  it('shows available languages and selects the default language', async () => {
```
Should this be wrapped in a describe that describes the initial state like ```describe('when the language chooser is first loaded')``` or something.

geoffreykwan commented on src/app/common/project-language-chooser/project-language-chooser.component.spec.ts:
```diff
@@ -0,0 +1,57 @@
+    expect(await options[0].isSelected()).toBeTrue();
+  });
+
+  it('keeps selected language option when language option changes', async () => {
```
Should this be wrapped in a describe that describes the action like ```describe('when the language options change')``` or something.

geoffreykwan reviewed (approved):
Looks good. I have some minor suggestions.

hirokiterashima committed:
Call projectLocale.getDefaultLocale() directly. Make hasLocale() private

hirokiterashima committed:
Clean up tests for improved readability
